src/cdk-experimental/scrollable-table-body/scrollable-table-body-layout.ts

```diff
@@ -0,0 +1,111 @@
+
+/**
+ * A {@link _TableLayoutStrategy} that enables scrollable body content for flex tables.
+ */
+@Injectable()
+class ScrollableTableBodyLayoutStrategy implements _TableLayoutStrategy {
+  private _pendingMaxHeight = 'none';
+  private _scrollViewport?: HTMLElement;
+  readonly headerCssClass = 'cdk-table-scrollable-table-header';
+  readonly bodyCssClass = 'cdk-table-scrollable-table-body';
+  readonly footerCssClass = 'cdk-table-scrollable-table-footer';
+
+  constructor(@Inject(DOCUMENT) private readonly _document: any) {
+  }
+
+  /**
+   * Returns the DOM structure for a native table. Scrollable body content is not supported for
+   * native tables. Return `null` to use the default {@link CdkTable} native table layout.
+   */
+  getNativeLayout(table: CdkTable<unknown>): _TableLayout {
+    return null;
+  }
+
+  /**
+   * Returns the DOM structure for a flex table with scrollable body content. Each row outlet
+   * (header, body, footer) is wrapped in a separate container. The specified max height is applied
+   * to the body row outlet to make its content scrollable.
+   */
+  getFlexLayout(table: CdkTable<unknown>): _TableLayout {
+    const documentFragment = this._document.createDocumentFragment();
+    const sections = [
+      {selector: this.headerCssClass, outlets: [table._headerRowOutlet]},
+      {selector: this.bodyCssClass, outlets: [table._rowOutlet, table._noDataRowOutlet]},
+      {selector: this.footerCssClass, outlets: [table._footerRowOutlet]},
+    ];
+
+    for (const section of sections) {
+      const element = this._document.createElement('div');
+      element.classList.add(section.selector);
+      for (const outlet of section.outlets) {
+        element.appendChild(outlet.elementRef.nativeElement);
+      }
+
+      documentFragment.appendChild(element);
+    }
+
+    this._scrollViewport = documentFragment.querySelector(`.${this.bodyCssClass}`);
+    this._scrollViewport!.style.overflow = 'auto';
+    this._applyMaxHeight(this._scrollViewport!, this._pendingMaxHeight);
+
+    return documentFragment;
+  }
+
+  /**
+   * Show a scroll bar if the table's body exceeds this height. The height may be specified with
+   * any valid CSS unit of measurement.
+   */
+  setMaxHeight(v: string) {
+    this._pendingMaxHeight = v;
+    if (this._scrollViewport) {
+      this._applyMaxHeight(this._scrollViewport, v);
+    }
+  }
+
+  private _applyMaxHeight(el: HTMLElement, maxHeight: string) {
+    el.style.maxHeight = maxHeight;
+  }
+}
+
+/** A directive that enables scrollable body content for flex tables. */
+@Directive({
+  selector: 'cdk-table[scrollableBody], mat-table[scrollableBody]',
+  providers: [
+    {provide: _TABLE_LAYOUT_STRATEGY, useClass: ScrollableTableBodyLayoutStrategy},
+  ]
+})
+export class CdkScrollableTableBody {
+  /**
+   * Show a scroll bar if the table's body exceeds this height. The height may be specified with
+   * any valid CSS unit of measurement.
+   */
+  @Input('scrollableBody')
```

Comment:
I struggled a bit to get this case working: I want my table to expand in height to fit the content of the container, and I don't want to have to set a `maxHeight`. Is that possible to do with this?

Reply:
I think the desired behavior will vary for each app:

1. Fixed height (e.g. `200px`, `60vh`, `calc(100vh - OFFSET_FROM_TOP)`)
2. Fill height of immediate parent element.
3. Fill remaining space in a higher level parent element (e.g immediate parent is a card and the table should fill the remaining space on the page).

The first use case is covered. IMO, the other two should be supported in another directive.

Reply:
What are the options for letting people set this via CSS? Turning it into a TS API makes it harder to e.g. use media queries, your Sass layout constants, etc.